SVSM utility scripts (#16)
amd/sev-utils: docs/svsm.md

# SVSM  Utility Script (Launch_SVSM.sh)

> This 'Launch_SVSM.sh'  utility script is a provisioning and test script that performs three tasks:

Reviewer: ```suggestion
'Launch_SVSM.sh' is a provisioning and testing script that performs three tasks:
```

Author: Agreed. Thanks for pointing that out; I'll update it.

> 1. Sets up an AMD EPYC CPU powered server by building host kernel, guest kernel, IGVM,  qemu, ovmf builds

Reviewer 2: ```suggestion
1. Sets up an AMD EPYC CPU powered server by building host kernel, guest kernel, IGVM, qemu, and ovmf builds
```

Author: Agreed.

> 1. Sets up an AMD EPYC CPU powered server by building host kernel, guest kernel, IGVM,  qemu, ovmf builds

Reviewer: ```suggestion
1. Sets up an AMD EPYC CPU powered server by building SVSM compatible host kernel, guest kernel, IGVM,  qemu and ovmf builds
```

Author: Agreed. Thanks for pointing that out; I'll update it.

> 2. Direct boot launch a SNP enabled guest with qemu using IGVM file

Reviewer: ```suggestion
2. Direct boot launch a SNP enabled guest with SVSM using QEMU and a IGVM file.
```

Author: How about like this "Launch an SNP-SVSM-enabled guest using SVSM and IGVM-supported options in the QEMU command"

> 3. Attest the SNP guest using the [virtee/snpguest](https://github.com/virtee/snpguest) 
> CLI tool.
> 
> 4. Verify measurement - igvmmeasure tool and snp attestation report

Reviewer: ```suggestion
3. Attest the SNP guest and validate the measurement using the [virtee/snpguest](https://github.com/virtee/snpguest) CLI tool and the igvmmeasure tool.
```

You did say 3 things and then enumerated 4

Author: Agreed. I will update.

Tested on the following OS distributions:
- Ubuntu 24.04

Image formats supported:
- qcow2

WARNING: 
This script installs developer packages on the system it is run on. 

> Beware and check 'install_dependencies' if there are any admin concerns.

Reviewer 2: We should essentially never assume it is okay to install dependencies on a machine. We should probably add in some logic to either ask a question or require a flag to install those dependencies. Opt-in instead of opt-out.

Author: Agreed. Thank you for pointing that out. It's important to ensure users are fully aware of what is being installed. 
 We can modify the script to prompt for confirmation before proceeding with the installation of dependencies. 
 This will give users the option to either proceed or opt-out. 
 I'll introduce logic that prompts users with a clear message.

[58207-using-sev-with-amd-epyc-processors.pdf](https://www.amd.com/content/dam/amd/en/documents/developer/58207-using-sev-with-amd-epyc-processors.pdf)

## Using the Script Utility

Download the script:
```
git clone https://github.com/AMDEPYC/Coconut-SVSM.git

```

Setup the host by building SVSM  patched versions of IGVM, qemu, ovmf and the host and guest linux kernels:
```
./Launch_SVSM.sh setup-host
```

A reboot will be necessary with SVSM enabled host kernel:
```
sudo reboot
```

When the system has finished rebooting back into the OS, launch a guest using 
the following command:
```
./Launch_SVSM.sh launch-guest
```

This will download a cloud-init ubuntu server jammy image that will be used as the 
guest disk. The guest is launched by passing qemu direct boot command line options 
for IGVM, initrd, kernel and the kernel append parameters.

Attest the guest using the following command:
```
./Launch_SVSM.sh attest-guest
```

The above result will show the contents of the SNP report and perform the 
report signature and certificate CA verification. It uses the igvmmeasure tool 
output from SVSM build to calculate the 

> expected launch measurement by measuring the IGVM file(SVSM,ovmf). This expected measurement 
> is then checked and verified against the launch measurement that is output from the 
> [virtee/snpguest](https://github.com/virtee/snpguest) tool. If the two measurements 
> match, then the test returns with a successful output.

Reviewer: ```suggestion
expected launch measurement by measuring the IGVM file(SVSM,ovmf). This expected measurement 
is then checked and verified against the launch measurement that is output from the 
Attestation Report. If the two measurements 
match, then the test returns with a successful output.
```

Author: Agreed. I will update.

## Stopping all Guests

All script created guests can be stopped by running the following command:
```
./Launch_SVSM.sh stop-guests
```

## BYO Image

> The SNP script utility provides support for the user to provide their own image.

Reviewer: ```suggestion
The SVSM script utility provides support for the user to provide their own image.
```

Author: Good catch. Agreed. I will update.
